tpe/plugin/src/EntityManagementFeatures.cc

```diff
@@ -0,0 +1,497 @@
+}
+
+/////////////////////////////////////////////////
+const std::string &EntityManagementFeatures::GetWorldName(
+  const Identity &_worldID) const
+{
+  static std::string worldName {};
+  auto it = this->worlds.find(_worldID);
+  if (it != this->worlds.end() && it->second != nullptr)
+  {
+    worldName = it->second->world->GetName();
+  }
+  return worldName;
+}
+
+/////////////////////////////////////////////////
+std::size_t EntityManagementFeatures::GetWorldIndex(
+  const Identity &_worldID) const
+{
+  // index should be 0 assuming there's only one world
+  auto it = this->worlds.find(_worldID);
+  if (it != this->worlds.end() && it->second != nullptr)
+    return std::distance(it, this->worlds.begin());
+  return -1;
+}
+
+/////////////////////////////////////////////////
+Identity EntityManagementFeatures::GetEngineOfWorld(
+  const Identity &) const
+{
+  return this->GenerateIdentity(0);
+}
+
+/////////////////////////////////////////////////
+std::size_t EntityManagementFeatures::GetModelCount(
+  const Identity &_worldID) const
+{
+  auto it = this->worlds.find(_worldID);
+  if (it != this->worlds.end() && it->second != nullptr)
+  {
+    return this->worlds.at(_worldID)->world->GetChildCount();
+  }
+  // return invalid index if world not found
+  return -1;
+}
+
+/////////////////////////////////////////////////
+Identity EntityManagementFeatures::GetModel(
+  const Identity &_worldID, const std::size_t _modelIndex) const
+{
+  if (this->worlds.find(_worldID) != this->worlds.end())
+  {
+    std::size_t modelId = this->indexInContainerToId(_worldID, _modelIndex);
+    auto it = this->models.find(modelId);
+    if (it != this->models.end() && it->second != nullptr)
+    {
+      return this->GenerateIdentity(it->first, it->second);
+    }
+  }
+  return this->GenerateInvalidId();
+}
+
+/////////////////////////////////////////////////
+Identity EntityManagementFeatures::GetModel(
+  const Identity &_worldID, const std::string &_modelName) const
+{
+  auto worldIt = this->worlds.find(_worldID);
+  if (worldIt != this->worlds.end() && worldIt->second != nullptr)
+  {
+    tpelib::Entity &modelEnt =
+      worldIt->second->world->GetChildByName(_modelName);
+    for (auto it = this->models.begin(); it != this->models.end(); ++it)
+    {
+      if (it->second != nullptr)
+      {
+        std::string name = it->second->model->GetName();
+        if (it->first == modelEnt.GetId() && name == modelEnt.GetName())
+        {
+          return this->GenerateIdentity(it->first, it->second);
+        }
+      }
+    }
+  }
+  return this->GenerateInvalidId();
+}
+
+/////////////////////////////////////////////////
+const std::string &EntityManagementFeatures::GetModelName(
+  const Identity &_modelID) const
+{
+  static std::string modelName {};
+  auto it = this->models.find(_modelID);
+  if (it != this->models.end() && it->second != nullptr)
+  {
+    modelName = it->second->model->GetName();
+  }
+  return modelName;
```

Comment:
I think we could use `ReferenceInterface` in this function to avoid lookups?

```
static std::string modelName =
    this->ReferenceInterface<ModelInfo>(_modelID)->model->GetName();
return modelName;
```

I now understand that `static std::string` is used because of the `const &` return type.  I think it's designed to match the return type of dart's [Skeleton::getName](https://dartsim.github.io/api/v5.1.4/d3/d19/classdart_1_1dynamics_1_1Skeleton.html#a583867ed1a17d3f2397562d68b860a7b) function